File: apps/web/lib/wechat.ts
/**
 * WeChat OAuth helpers for the Chinese market
 *
 * Documentation: https://developers.weixin.qq.com/doc/oplatform/Website_App/WeChat_Login/Wechat_Login.html
 */

const WECHAT_APP_ID = process.env.WECHAT_APP_ID || ''
const WECHAT_APP_SECRET = process.env.WECHAT_APP_SECRET || ''
const WECHAT_REDIRECT_URI = process.env.NEXT_PUBLIC_WECHAT_REDIRECT_URI || ''

// WeChat OAuth URLs
const WECHAT_AUTHORIZE_URL = 'https://open.weixin.qq.com/connect/qrconnect'
const WECHAT_ACCESS_TOKEN_URL = 'https://api.weixin.qq.com/sns/oauth2/access_token'
const WECHAT_USERINFO_URL = 'https://api.weixin.qq.com/sns/userinfo'

export interface WeChatTokenResponse {
  access_token: string
  expires_in: number
  refresh_token: string
  openid: string
  scope: string
  unionid?: string
  errcode?: number
  errmsg?: string
}

export interface WeChatUserInfo {
  openid: string
  nickname: string
  sex: number
  province: string
  city: string
  country: string
  headimgurl: string
  privilege: string[]
  unionid?: string
  errcode?: number
  errmsg?: string
}

/**
 * Generate WeChat OAuth authorization URL
 * This URL should be used to redirect users to WeChat for login
 */
export function getWeChatAuthUrl(state?: string): string {
  const params = new URLSearchParams({
    appid: WECHAT_APP_ID,
    redirect_uri: WECHAT_REDIRECT_URI,
    response_type: 'code',
    scope: 'snsapi_login',
    state: state || generateState(),
  })

  return `${WECHAT_AUTHORIZE_URL}?${params.toString()}#wechat_redirect`
}

/**
 * Exchange authorization code for access token
 */
export async function getWeChatAccessToken(code: string): Promise<WeChatTokenResponse> {
  const params = new URLSearchParams({
    appid: WECHAT_APP_ID,
    secret: WECHAT_APP_SECRET,
    code,
    grant_type: 'authorization_code',
  })

  const response = await fetch(`${WECHAT_ACCESS_TOKEN_URL}?${params.toString()}`)
  const data = await response.json()

  if (data.errcode) {
    throw new Error(`WeChat OAuth error: ${data.errmsg} (${data.errcode})`)
  }

  return data
}

/**
 * Get user info using access token
 */
export async function getWeChatUserInfo(accessToken: string, openid: string): Promise<WeChatUserInfo> {
  const params = new URLSearchParams({
    access_token: accessToken,
    openid,
    lang: 'zh_CN',
  })

  const response = await fetch(`${WECHAT_USERINFO_URL}?${params.toString()}`)
  const data = await response.json()

  if (data.errcode) {
    throw new Error(`WeChat user info error: ${data.errmsg} (${data.errcode})`)
  }

  return data
}

/**
 * Generate a random state parameter for CSRF protection
 */
export function generateState(): string {
  return Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15)
}

/**
 * Verify state parameter to prevent CSRF attacks
 */
export function verifyState(receivedState: string, storedState: string): boolean {
  return receivedState === storedState
}

/**
 * Get WeChat QR code image URL for scanning
 * This is for scenarios where you want to display the QR code directly
 */
export function getWeChatQRCodeUrl(state?: string): string {
  const authUrl = getWeChatAuthUrl(state)
  // The authUrl itself will display a QR code when opened in a browser
  return authUrl
}

/**
 * Check if WeChat is configured
 */
export function isWeChatConfigured(): boolean {
  return Boolean(WECHAT_APP_ID && WECHAT_APP_SECRET && WECHAT_REDIRECT_URI)
}

File: apps/api/app/services/wechat.py
"""
WeChat OAuth service for candidate authentication.

Supports both WeChat Web (PC) and WeChat Mini Program login.
"""
import httpx
from typing import Optional
from dataclasses import dataclass
from ..config import settings


@dataclass
class WeChatUserInfo:
    """User info returned from WeChat OAuth."""
    openid: str
    unionid: Optional[str] = None
    nickname: Optional[str] = None
    headimgurl: Optional[str] = None
    sex: Optional[int] = None
    province: Optional[str] = None
    city: Optional[str] = None
    country: Optional[str] = None


class WeChatOAuthService:
    """
    WeChat OAuth 2.0 authentication service.

    Flow:
    1. Frontend redirects user to WeChat authorization URL
    2. User authorizes, WeChat redirects back with code
    3. Backend exchanges code for access_token and openid
    4. Backend fetches user info (if scope includes snsapi_userinfo)
    5. Backend creates/updates candidate and returns JWT

    References:
    - Web OAuth: https://developers.weixin.qq.com/doc/oplatform/Website_App/WeChat_Login/Wechat_Login.html
    - Mini Program: https://developers.weixin.qq.com/miniprogram/dev/framework/open-ability/login.html
    """

    # WeChat API endpoints
    WEB_AUTH_URL = "https://open.weixin.qq.com/connect/qrconnect"
    ACCESS_TOKEN_URL = "https://api.weixin.qq.com/sns/oauth2/access_token"
    USERINFO_URL = "https://api.weixin.qq.com/sns/userinfo"
    MINI_PROGRAM_CODE2SESSION_URL = "https://api.weixin.qq.com/sns/jscode2session"

    def __init__(self):
        self.app_id = settings.wechat_app_id
        self.app_secret = settings.wechat_app_secret

    def is_configured(self) -> bool:
        """Check if WeChat OAuth is properly configured."""
        return bool(self.app_id and self.app_secret)

    def get_web_auth_url(
        self,
        redirect_uri: str,
        state: str = "",
        scope: str = "snsapi_login"
    ) -> str:
        """
        Generate WeChat Web OAuth authorization URL.

        Args:
            redirect_uri: URL to redirect after authorization (must be URL-encoded)
            state: Optional state parameter for CSRF protection
            scope: OAuth scope (snsapi_login for web)

        Returns:
            Authorization URL to redirect the user to
        """
        import urllib.parse

        params = {
            "appid": self.app_id,
            "redirect_uri": redirect_uri,
            "response_type": "code",
            "scope": scope,
            "state": state,
        }

        query_string = urllib.parse.urlencode(params)
        return f"{self.WEB_AUTH_URL}?{query_string}#wechat_redirect"

    async def exchange_code_for_token(self, code: str) -> dict:
        """
        Exchange authorization code for access token.

        Args:
            code: Authorization code from WeChat callback

        Returns:
            Dict with access_token, refresh_token, openid, etc.

        Raises:
            Exception if the request fails or returns an error
        """
        async with httpx.AsyncClient(timeout=10.0) as client:
            response = await client.get(
                self.ACCESS_TOKEN_URL,
                params={
                    "appid": self.app_id,
                    "secret": self.app_secret,
                    "code": code,
                    "grant_type": "authorization_code",
                }
            )
            response.raise_for_status()
            data = response.json()

            if "errcode" in data and data["errcode"] != 0:
                raise Exception(f"WeChat OAuth error: {data.get('errmsg', 'Unknown error')}")

            return data

    async def get_user_info(self, access_token: str, openid: str) -> WeChatUserInfo:
        """
        Fetch user info from WeChat.

        Args:
            access_token: Valid access token
            openid: User's OpenID

        Returns:
            WeChatUserInfo with user details
        """
        async with httpx.AsyncClient(timeout=10.0) as client:
            response = await client.get(
                self.USERINFO_URL,
                params={
                    "access_token": access_token,
                    "openid": openid,
                    "lang": "zh_CN",
                }
            )
            response.raise_for_status()
            data = response.json()

            if "errcode" in data and data["errcode"] != 0:
                raise Exception(f"WeChat userinfo error: {data.get('errmsg', 'Unknown error')}")

            return WeChatUserInfo(
                openid=data["openid"],
                unionid=data.get("unionid"),
                nickname=data.get("nickname"),
                headimgurl=data.get("headimgurl"),
                sex=data.get("sex"),
                province=data.get("province"),
                city=data.get("city"),
                country=data.get("country"),
            )

    async def mini_program_login(self, js_code: str) -> dict:
        """
        WeChat Mini Program login using js_code.

        Args:
            js_code: Login code from wx.login()

        Returns:
            Dict with session_key, openid, unionid (if available)
        """
        async with httpx.AsyncClient(timeout=10.0) as client:
            response = await client.get(
                self.MINI_PROGRAM_CODE2SESSION_URL,
                params={
                    "appid": self.app_id,
                    "secret": self.app_secret,
                    "js_code": js_code,
                    "grant_type": "authorization_code",
                }
            )
            response.raise_for_status()
            data = response.json()

            if "errcode" in data and data["errcode"] != 0:
                raise Exception(f"Mini Program login error: {data.get('errmsg', 'Unknown error')}")

            return data

    async def authenticate(self, code: str, is_mini_program: bool = False) -> WeChatUserInfo:
        """
        Complete OAuth flow: exchange code and get user info.

        Args:
            code: Authorization code or Mini Program js_code
            is_mini_program: Whether this is a Mini Program login

        Returns:
            WeChatUserInfo with user details (may only have openid for Mini Program)
        """
        if is_mini_program:
            # Mini Program flow - only returns openid and session_key
            data = await self.mini_program_login(code)
            return WeChatUserInfo(
                openid=data["openid"],
                unionid=data.get("unionid"),
            )
        else:
            # Web OAuth flow - can get full user info
            token_data = await self.exchange_code_for_token(code)
            return await self.get_user_info(
                access_token=token_data["access_token"],
                openid=token_data["openid"],
            )


# Global instance
wechat_service = WeChatOAuthService()
